src/python/pants/backend/go/util_rules/external_module.py

```diff
@@ -34,6 +35,86 @@
 from pants.util.strutil import strip_v2_chroot_path
 
 
+@dataclass(frozen=True)
+class DownloadedExternalModules:
+    """All downloaded modules, along with the input `go.mod` and `go.sum`."""
+
+    digest: Digest
+
+
+@dataclass(frozen=True)
+class DownloadExternalModulesRequest:
+    """Download all modules from the `go.mod`.
+
+    The `go.mod` and `go.sum` must already be up-to-date.
+    """
+
+    go_mod_stripped_digest: Digest
+
+
+@rule
+async def download_external_modules(
+    request: DownloadExternalModulesRequest,
+) -> DownloadedExternalModules:
+    download_result = await Get(
+        ProcessResult,
+        GoSdkProcess(
+            command=("mod", "download", "-json", "all"),
+            input_digest=request.go_mod_stripped_digest,
+            # TODO: make this more descriptive: point to the actual `go_mod` target or path.
+            description="Download all external Go modules",
+            output_files=("go.mod", "go.sum"),
+            output_directories=("gopath",),
+        ),
+    )
+
+    # Check that the root `go.mod` and `go.sum` did not change.
+    result_go_mod_digest = await Get(
+        Digest, DigestSubset(download_result.output_digest, PathGlobs(["go.mod", "go.sum"]))
+    )
+    if result_go_mod_digest != request.go_mod_stripped_digest:
+        # TODO: make this a more informative error.
+        contents = await Get(DigestContents, Digest, result_go_mod_digest)
+
+        raise Exception(
+            "`go.mod` and/or `go.sum` changed! Please run `go mod tidy`.\n\n"
+            f"{contents[0].content.decode()}\n\n"
+            f"{contents[1].content.decode()}\n\n"
+        )
+
+    # TODO: strip `gopath` and other paths?
```

Comment:
FYI the directory structure for the module cache under GOPATH is well-defined so you probably don't need to strip paths, just figure out what you want in or out of the Digest.